src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java

```diff
@@ -202,6 +203,407 @@ public void testConfigSettingVisibilityAlwaysCheckedAtUse() throws Exception {
     assertThat(hasErrors(getConfiguredTarget("//:my_target"))).isFalse();
   }
 
+  @Test
+  public void testAspectImplicitDependencyCheckedAtDefinition_visible() throws Exception {
+    scratch.file("inner_aspect/BUILD");
+    scratch.file(
+        "inner_aspect/lib.bzl",
+        "InnerAspectInfo = provider()",
+        "def _impl_inner_aspect(ctx, target):",
+        "  return [InnerAspectInfo()]",
+        "inner_aspect = aspect(",
+        "  _impl_inner_aspect,",
+        "  attrs = { '_inner_aspect_tool': attr.label(default = '//tools:inner_aspect_tool') },",
+        "  provides = [InnerAspectInfo],",
+        ")");
+
+    scratch.file("outer_aspect/BUILD");
+    scratch.file(
+        "outer_aspect/lib.bzl",
+        "load('//inner_aspect:lib.bzl', 'InnerAspectInfo')",
+        "def _impl_outer_aspect(ctx, target):",
+        "  return []",
+        "outer_aspect = aspect(",
+        "  _impl_outer_aspect,",
+        "  attrs = { '_outer_aspect_tool': attr.label(default = '//tools:outer_aspect_tool') },",
+        "  required_aspect_providers = [InnerAspectInfo],",
+        ")");
+
+    scratch.file("rule/BUILD");
+    scratch.file(
+        "rule/lib.bzl",
+        "load('//inner_aspect:lib.bzl', 'inner_aspect')",
+        "load('//outer_aspect:lib.bzl', 'outer_aspect')",
+        "def _impl(ctx):",
+        "  pass",
+        "my_rule = rule(",
+        "  _impl,",
+        "  attrs = {",
+        "    'dep': attr.label(aspects = [inner_aspect, outer_aspect]),",
+        "    '_rule_tool': attr.label(default = '//tools:rule_tool'),",
+        "  },",
+        ")",
+        "simple_starlark_rule = rule(",
+        "  _impl,",
+        ")");
+
+    scratch.file(
+        "foo/BUILD",
+        "load('//rule:lib.bzl','my_rule', 'simple_starlark_rule')",
+        "simple_starlark_rule(name = 'simple_dep')",
+        "my_rule(name = 'target_with_aspects', dep = ':simple_dep')");
+
+    scratch.file(
+        "tools/BUILD",
+        "sh_binary(",
+        "  name = 'outer_aspect_tool',",
+        "  srcs = ['a.sh'],",
+        "  visibility = ['//outer_aspect:__pkg__'],",
+        ")",
+        "sh_binary(",
+        "  name = 'inner_aspect_tool',",
+        "  srcs = ['a.sh'],",
+        "  visibility = ['//inner_aspect:__pkg__'],",
+        ")",
+        "sh_binary(",
+        "  name = 'rule_tool',",
+        "  srcs = ['a.sh'],",
+        "  visibility = ['//rule:__pkg__'],",
+        ")");
+
+    useConfiguration("--incompatible_visibility_private_attributes_at_definition");
+
+    update("//foo:target_with_aspects");
+    assertThat(hasErrors(getConfiguredTarget("//foo:target_with_aspects"))).isFalse();
```

Comment:
This should be:
```
scratch
useConfiguration // arrange

update // act

assertThat // assert
```

Same in `testAspectImplicitDependencyCheckedAtDefinition_visibleWithNameCollision`